### PR TITLE
Add rails 7.2 support

### PR DESCRIPTION
**Pull request by zjohl**
This PR adds Rails 7.2 support including backwards compatibility for Rails 7.1.

We made the follow changes to support Rails 7.1:

1. New logic for registering adapters ([Rails commit](https://github.com/rails/rails/commit/009c7e74117690f0dbe200188a929b345c9306c1)) 

The majority of the required changes here are due to a change to way we register new adapters. In the past, this was by defining a `odbc_connection` function that could return the correct adapter based on the provided connection. This would allow us to use a single type of adapter (`odbc`) and return a specific subclass for the provided database.

This [has been replaced](https://github.com/rails/rails/commit/009c7e74117690f0dbe200188a929b345c9306c1) by a `register` function that our gem must call to register a specific class as our adapter. This means our initializer will always get called and we can't (easily) return a subclass from the initializer.

We now register 3 adapters for this gem. The new ones are the `postgres_odbc` and `mysql_odbc` adapters which have database-specific logic. The `odbc` adapater is now specifically referring to the `NullODBCAdapter` which is the more generic implementation 

2. Don't require an active connection for table and column quoting  ([Rails commit](https://github.com/rails/rails/commit/0016280f4fde55d96738887093dc333aae0d107b))

`quote_column_name` is used to quote column names and was previously an instance method which allowed us to check the active connection to figure out how to handle quoting. This is now a class method which means each adapter must know how to handle column quoting without an actual connection. I've largely leaned on the logic used by the Rails Mysql and Postgresql adapters which meant that we now need to rely on the PG gem.

### CODE REVIEW

File: lib/odbc_adapter/database_statements.rb [hunk 1]
```diff
@@ -27,7 +27,7 @@ def insert(arel, name = nil, pk = nil, id_value = nil, sequence_name = nil, bind
     # Executes +sql+ statement in the context of this connection using
     # +binds+ as the bind substitutes. +name+ is logged along with
     # the executed +sql+ statement.
-    def internal_exec_query(sql, name = 'SQL', binds = [], prepare: false) # rubocop:disable Lint/UnusedMethodArgument
```
Comment (zjohl) [inline lib/odbc_adapter/database_statements.rb:30]: Function args changed here: https://github.com/rails/rails/commit/eabcff22a86213521eecafbf2dbf3ad82714f941#diff-88e231c48eb5559ac3be58a024b019e86f7c74d388f5d49fac3f4e47bf851cc6R516

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -2,7 +2,7 @@ name: Ruby
 
 on:
   push:
-    branches: [ master ]
+    # branches: [ master ]
```
Comment (kevin-j-m) [inline .github/workflows/test.yml:5]: We should back out this change.

File: Gemfile [hunk 3]
```diff
@@ -2,4 +2,5 @@ source 'https://rubygems.org'
 
 gemspec
 
-gem 'activerecord', '>= 7.1.0'
+gem 'activerecord', '~> 7.2.2'
```
Comment (kevin-j-m) [inline Gemfile:5]: If we're still compatible with 7.1, do we need this change? This will prevent us from using it with 7.1.

File: lib/active_record/connection_adapters/odbc_adapter.rb [hunk 4]
```diff
@@ -50,15 +51,20 @@ class ODBCAdapter < AbstractAdapter
 
       def initialize(config_or_deprecated_connection, deprecated_logger = nil, deprecated_connection_options = nil, deprecated_config = nil, database_metadata = nil)
         super(config_or_deprecated_connection, deprecated_logger, deprecated_connection_options, deprecated_config)
-        @raw_connection = config_or_deprecated_connection
-
-        if database_metadata
-          @database_metadata = database_metadata
+        @config = deprecated_config
```
Comment (kevin-j-m) [inline lib/active_record/connection_adapters/odbc_adapter.rb:54]: It appears that in either case of the if conditional, we then reset `@config` to something. Do we still need this?

File: lib/active_record/connection_adapters/odbc_adapter.rb [hunk 5]
```diff
@@ -50,15 +51,20 @@ class ODBCAdapter < AbstractAdapter
...
+        if config_or_deprecated_connection.try(:get_info, ODBC.const_get("SQL_DBMS_NAME"))
+          @raw_connection = config_or_deprecated_connection
+          @config = deprecated_config
+          configure_time_options(@raw_connection)
         else
-          @database_metadata = ::ODBCAdapter::DatabaseMetadata.new(@raw_connection)
+          config = config_or_deprecated_connection
+          setup = ::ODBCAdapter::ConnectionSetup.new(config.symbolize_keys)
+          setup.build
+          @config = setup.config
```
Comment (kevin-j-m) [inline lib/active_record/connection_adapters/odbc_adapter.rb:63]: Why do we not need the time options on this side of the if test?

File: lib/odbc_adapter/adapters/null_odbc_adapter.rb [hunk 6]
```diff
@@ -4,15 +4,36 @@ module Adapters
...
+        quote_char = identifier_quote_char.to_s.strip
+
+        return name if quote_char.length.zero?
+        quote_char = quote_char[0]
+
+        # Avoid quoting any already quoted name
+        return name if name[0] == quote_char && name[-1] == quote_char
+
+        # If upcase identifiers, only quote mixed case names.
+        if upcase_identifiers?
```
Comment (kevin-j-m) [inline lib/odbc_adapter/adapters/null_odbc_adapter.rb:24]: If this is always true, why do we need this? I recognize that this is the piece of the logic that previously relied on instance state of the connection.

File: lib/odbc_adapter/database_statements.rb (hunk 1 above)
Comment (kevin-j-m) [reply lib/odbc_adapter/database_statements.rb:30]: oooOOOooo, I think I missed this.

**Review by kevin-j-m — COMMENTED** [review]
Exciting!

File: lib/active_record/connection_adapters/odbc_adapter.rb (hunk 5 above)
Comment (zjohl) [reply lib/active_record/connection_adapters/odbc_adapter.rb:63]: `connect` (on the next line) calls `configure_time_options`

File: lib/active_record/connection_adapters/odbc_adapter.rb (hunk 4 above)
Comment (zjohl) [reply lib/active_record/connection_adapters/odbc_adapter.rb:54]: No, good catch!

File: Gemfile (hunk 3 above)
Comment (zjohl) [reply Gemfile:5]: I don't think so, I'll unroll this

File: lib/odbc_adapter/adapters/null_odbc_adapter.rb (hunk 6 above)
Comment (zjohl) [reply lib/odbc_adapter/adapters/null_odbc_adapter.rb:24]: We don't need this if it's always true, I am hoping to test this with Snowflake and confirm that this is working as expected and then I can definitely clean this up a bit more

I'm thinking it also makes sense to add some tests around using the null adapter to query postgres (assuming that works) as a sanity check that this adapter isn't totally broken

Comment (kevin-j-m) [reply lib/odbc_adapter/adapters/null_odbc_adapter.rb:24]: > I'm thinking it also makes sense to add some tests around using the null adapter to query postgres (assuming that works) as a sanity check that this adapter isn't totally broken

Sounds good.

File: test/generic_adapter_test.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
+class GenericAdapterTest < Minitest::Test
```
Comment (zjohl) [inline test/generic_adapter_test.rb:1]: Super lame tests to just confirm we can use the NullODBCAdapter to do postgres operations

If we do run into issues in production with the adapter, this might end up being the right place to add additional tests confirming that we've fixed any specific issues

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -2,7 +2,7 @@ name: Ruby
 
 on:
   push:
-    branches: [ master ]
+    branches: [ main ]
```
Comment (zjohl) [inline .github/workflows/test.yml:5]: I renamed `master` to `main` so this change is just reflecting that update

File: lib/odbc_adapter/adapters/null_odbc_adapter.rb (hunk 6 above)
Comment (kevin-j-m) [reply lib/odbc_adapter/adapters/null_odbc_adapter.rb:24]: We now have the tests (well done, and thank you very much). Have we checked against snowflake? Are we going to remove this conditional? Or retain it and modify the implementation of `upcase_identifiers?`?

**Review by kevin-j-m — APPROVED** [review]
This is very exciting. Great work on this. Thank you.

File: lib/odbc_adapter/adapters/null_odbc_adapter.rb (hunk 6 above)
Comment (zjohl) [reply lib/odbc_adapter/adapters/null_odbc_adapter.rb:24]: Oof, thanks for the reminder. I did try this against Snowflake and queried without any issues

Double checking the snowflake docs, I think this is the right configuration so I'll clean up this code a bit
